src/util/psbt/mod.rs

```diff
@@ -208,6 +217,471 @@ impl PartiallySignedTransaction {
+            return Ok(OutputType::Bare);
+        }
+
+        if spk.is_v0_p2wpkh() {
+            return Ok(OutputType::Wpkh);
+        }
+
+        if spk.is_v0_p2wsh() {
+            return Ok(OutputType::Wsh);
+        }
+
+        if spk.is_p2sh() {
+            if input.redeem_script.as_ref().map(|s| s.is_v0_p2wpkh()).unwrap_or(false) {
+                return Ok(OutputType::ShWpkh);
+            }
+            if input.redeem_script.as_ref().map(|x| x.is_v0_p2wsh()).unwrap_or(false) {
+                return Ok(OutputType::ShWsh);
+            }
+            return Ok(OutputType::Sh);
+        }
+
+        if spk.is_v1_p2tr() {
+            return Ok(OutputType::Tr);
+        }
+
+        // Something is wrong with the input scriptPubkey or we do not know how to sign
+        // because there has been a new softfork that we do not yet support.
+        Err(SignError::UnknownOutputType)
+    }
+}
+
+/// Key material for signing with either ECDSA or schnorr.
+#[derive(Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))] // FIXME: How do we debug in no-std environment?
+pub enum PrivateKeyOrPair {
+    /// Private key used for ECDSA signing.
+    Key(PrivateKey),
+    /// Key pair used for schnorr signing.
+    Pair(KeyPair),
+}
+
+impl From<PrivateKey> for PrivateKeyOrPair {
+    fn from(k: PrivateKey) -> Self {
+        PrivateKeyOrPair::Key(k)
+    }
+}
+
+impl From<KeyPair> for PrivateKeyOrPair {
+    fn from(k: KeyPair) -> Self {
+        PrivateKeyOrPair::Pair(k)
+    }
+}
+
+/// Data required to call [`GetKey`] to get the private key to sign an input.
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum KeyRequest {
+    /// Request a private key using the associated public key.
+    Pubkey(PublicKey),
+    /// Request a private key using BIP-32 fingerprint and derivation path.
+    Bip32(KeySource),
+}
+
+/// Trait to get a private key from a key request, key is then used to sign an input.
+pub trait GetKey {
+    /// An error occurred while getting the key.
+    type Error: core::fmt::Debug;
+
+    /// Attempts to get the private key for `key_request`.
+    ///
+    /// # Returns
+    /// - `Some(key)` if the key is found.
+    /// - `None` if the key was not found but no error was encountered.
+    /// - `Err` if an error was encountered while looking for the key.
+    fn get_key<C: Signing>(&self, key_request: KeyRequest, secp: Option<&Secp256k1<C>>) -> Result<Option<PrivateKey>, Self::Error>;
+}
+
+impl GetKey for ExtendedPrivKey {
+    type Error = GetKeyError;
+
+    fn get_key<C: Signing>(&self, key_request: KeyRequest, secp: Option<&Secp256k1<C>>) -> Result<Option<PrivateKey>, Self::Error> {
+        let secp = secp.ok_or(GetKeyError::RequireSecp)?;
+
+        match key_request {
+            KeyRequest::Pubkey(_) => Err(GetKeyError::NotSupported),
+            KeyRequest::Bip32((_fingerprint, path)) => {
+                let k = self.derive_priv(secp, &path)?;
```

Comment:
We should check whether the fingerprint matches here or not. 
```rust
if self.fingerprint() == fingerprint { 
   // derive key here
} else {
    None
}
```

In the `HashSet<ExtendedPrivKey>`, we can then iterate over all the `ExtendedPrivKey`s and derive only from those whose fingerprint is matched.

Reply:
The `HashSet<ExtendedPrivKey>` is the closest that we can get to a real wallet with multiple accounts etc.

Reply:
Cool, thanks man.